### DISCUSSION TITLE
PermissionManager User Permissions and Roles Long Exhaustive List

### DISCUSSION THREAD

**Discussion by exorgroup** [Laravel-Backpack/community-forum#752] [General]
I have implemented this package in the project- thanks for saving some good time

This project may have over 30 tables and each table will require CRUD and print option as well.  How everything is set up this means that there will be a very long list of permissions that can get users confused - example country_create, country_edit, country_delete, country_view, country_print and then town_create, town_edit, town_delete, town_view, town_print etc etc etc

Was wondering if there is a possibility of grouping these together so at least the user can see like a heading Country and then the roles attributed to that - ideally having like a collapsible setup so they do not need to scroll through 150 rows or more

Simple way of grouping could be using the underscore to pipe to delineate the groups like setup_country_create, setup_town create and finance_invoice_create.  This would imply a structure as following

Setup >
* Country >
   * Create
* Town >
  * Create
  
Finance >
* Invoice >
  * Create

Thanks

**Comment by pxpm** [comment]
Hey there @exorgroup , thanks for the suggestion. At the moment we don't have nothing like that, but nothing stopping you from building a field like that yourself 🙏 

It's an interesting use-case tho, so maybe more people would be interested in it. we are currently running a poll where users can suggest and vote for the next features we will build, so maybe you can suggest it there too: https://github.com/Laravel-Backpack/community-forum/discussions/751

If more people are interested in something like this we may give it the proper priority. 

You can add this, or any suggestion you see fit there. 

Cheers

**Reply by exorgroup** [reply]
Do not think a field would do it as field is declared 
```php
CRUD::column([
            'name' => 'ISOCode',
            'type' => 'text',
            'label' => 'ISO Code',
          ]);
```

it needs to be something like 
```php
CRUD::group([
            'name' => 'Country',
            'type' => 'text',
            'label' => 'Country',
          ]);


CRUD::group([
            'name' => 'Locality',
            'type' => 'text',
            'label' => 'Locality',
          ]);
```

Every time you add a group it means that it is a group of the one of the above

**Reply by pxpm** [reply]
I am not sure, that's creating "two independant groups". 
You want a field to have "awareness" of those groups and how they are stored in database (only parent_id ? in a nested set pattern?) 
Similar to how we did with `checklist_dependency`, but instead of having only "one direct dependency" you need to orchestrate that for multiple dependencies in more levels. 

In my opinion I would do that in a single field, that will probably send an array of something to the backend. The front-end needs the "tree structure", you can use a bit of javascript to show/hide the elements when clicked, and store the selected ones in some array that will be sent. 

So something more along the lines of:
```php
CRUD::field('custom_tree_selection')
    ->type('my_custom_dependency_field')
    ->groups([
        'groups1' => ....,
    ])
```

**Reply by exorgroup** [reply]
@pxpm 
I have modified the blade template checklist_dependency and created checklist_dependency_with_grouping and now I have this result with bootstrap accordion implemented in the user section

![permissions](https://github.com/Laravel-Backpack/community-forum/assets/56255746/71251c29-4616-4f6f-ab89-cd49de3e38c6)

It works by finding the first underscore and splits on it to create the header - the permissions are still unchanged so the implementation in the system remains the same 

I need some tips in implementing the same approach using on the Edit and Create Permissions field.  They implement using checklist so my assumption is that I need to modify checklist.blade.php right?

**Reply by pxpm** [reply]
Yes, you will probably need a custom `checklist_with_grouping`. 

It's starting to gain shape. 👀 
Why do you repeat the navigation so much times ? Apart from that is looks very nice. 👌 

If you want later you can publish it as a addon package for Backpack that other developers can re-use. 🙏

**Reply by exorgroup** [reply]
@pxpm thanks for the compliments 

the repeated navigation > initially I was opting for the list at the very top with a link to scroll to top with each section so that one can easily scroll to the top and click the next link (basically 2 actions) - I was experimenting than with the idea of just clicking on the navigation directly and jump from place to another without the need of scrolling to top first (basically 1 action) - I do agree that it may be a bit overwhelming but sort of liked it, even though aesthetically may not the best

publish as add on package - it is something that would consider as I think it may help other developers especially the checklist_with_grouping

Roles > this is how it looks - it does not look bad like this when the accordion is collapsed (initial stage)
![roles00](https://github.com/Laravel-Backpack/community-forum/assets/56255746/e6d867d1-5fdd-4969-87ba-869660d17ca7)

this is how it looks when you start opening accordion tabs - currently are set not to collapse the opened ones so it is quickly noticed the one altered

![roles01](https://github.com/Laravel-Backpack/community-forum/assets/56255746/010587a9-9f63-427b-9ed9-e044c433bf21)
 and 
![roles02](https://github.com/Laravel-Backpack/community-forum/assets/56255746/228963ee-be6f-494c-ac17-63d333532371)

**Reply by pxpm** [reply]
Indeed they look nice, but as you start opening the tabs I feel the page get overwhelmed with too much information/repetition.

How about you make the `quick navigation` a sticky header ? That way when you scroll down, it will scroll down with you and you have it always available. Similar to how the table columns behave on the list operation. If you load 100 entries and scroll down, the table header will keep up with you to help you read the table. 

Cheers

**Reply by exorgroup** [reply]
@pxpm - point noted on sticky header

I have done the change and like it - it looks an elegant solution

![roles03](https://github.com/Laravel-Backpack/community-forum/assets/56255746/6484ad37-da01-4e1a-aa65-4b3a9c23ff86)

What do you think?

Current position is scrolled up to the top

**Reply by exorgroup** [reply]
@pxpm I have updated a bit more and I think it may be a better design option - it has highlight on the row and icons for create/add/new, update/edit, delete, read/view and print - what do you think?

![newPermissions](https://github.com/Laravel-Backpack/community-forum/assets/56255746/c9835215-8b8c-4bba-87b1-e9c09007e285)
![newRoles01](https://github.com/Laravel-Backpack/community-forum/assets/56255746/20e046c5-6b10-4fa5-a62b-0e34738a5a94)

**Reply by pxpm** [reply]
ok ... this is starting to get spicy. 🌶️ 🤓 

I have mixed feelings from the icons. From one side, I like it, I like "button states", I think they "clean up" the UI a lot. 
But my "devil advocate" says me that this is not as "accessible" as a check box with a text (or a switch). 
For example it makes more difficult to assert which ones are active from the ones that are inactive when only colors change.
It makes it completely unusable for color impaired people, and it is only "mental recognizable" for this "common" use cases, any other icons out of the "ordinary create, edit, preview etc" start to need docs alongside them. 

If I would change checkbox for buttons, I would still make something show up on button indicating that's active, similar to the checkbox, and with the permission text, something like:
![image](https://github.com/Laravel-Backpack/community-forum/assets/7188159/57776647-b7e2-44d7-9686-765ead6e9173)

I would also probably remove the "Permissions" text on top of the "quick links" since you get it again down in the table. 

The a,b,c,d,e etc also looks a little bit out of place, I would either center it or make it full width. It would probably look a little bit better without the top and bottom borders too. 

While on it, I would probably change "quick links" for "FILTER:".   "FILTER: A | B | C | D | E .. "

This are my personal views and opinions, it's totally fine to have a completely different opinion than mine and still stand correct. There is nothing stopping you from supporting "multiple point of views" in your project, or supporting only yours. It's totally up to you since you are the one putting in the effort 💪 

You are doing a good job, keep it up, ping me if you need my help with something or just to give a feedback 🙏

**Reply by exorgroup** [reply]
@pxpm 

State > I agree with your argument that it makes a bit difficult to assert - will work to have some other visual clue as suggested

Quick links to filter - I agree that quick links might not be the right words to use but I have reservations to the use of the word filter as well as when a filter is clicked than it is assumed that the list is reduced in some way - this just jumps/scrolls to that location - potentially need to hit the thesaurus for a better word to jump 

What I would like some help on is to introduce some parameters for example FilterText (gives the user the option whatever text they want instead of Quick Links), TextOnly to have buttons showing only text etc.  Can you provide me some help how that can be achieved please?

Thanks

**Reply by pxpm** [reply]
Maybe `Navigation` then ? I thought you were about to implement some ajax or even js filtering, so that I click on "A", and only A started entries are show.  It would also be good to identify differently the letters that has no "functions" associated. 

Probably the more straightforward way is by creating JS objects of the permissions/functions, and then use some sort of filtering on the javascript collections when the user writes on an input. This does not required an ajax endpoint and since you already have all the information on page you can just manipulate it with javascript. 

About the configs, you can have a key on your field definition that allow developer to setup some configs like: 
```php
CRUD::field('my_custom_checklist')
    ->type('advanced_checklist')
    ->uiOptions([
        'textOnly' => true, 
        'withIcon' => false,
// .. etc
```

Cheers

**Reply by exorgroup** [reply]
@pxpm
I have updated again the UI based your feedback as points as they made sense to me - I am also using the people list example in tabler.io for the scroller.  Let me know you thoughts about it please 

![roles](https://github.com/Laravel-Backpack/community-forum/assets/56255746/0d62769c-e048-4b4d-933a-521d48fd884e)

**Comment by exorgroup** [comment]
@pxpm 
it has been a while since I developed this.  I am working on new project and wanted to share the updated versions.  Open to comments - might remove the icons and replace with text only on a particular version. Why one click there are 2 visual elements indicating the status - the colour and the check mark both change based on state

![image](https://github.com/user-attachments/assets/9ab98b7a-b808-4256-94d4-f0c3c9fe0541)

![image](https://github.com/user-attachments/assets/11f38c5b-d136-4498-ac92-698154ce21a8)
